### DISCUSSION TITLE
Lesson 9: testPerformUpkeepRevertsIfCheckUpkeepIsFalse

### DISCUSSION THREAD

**Discussion by codeplay22** [Cyfrin/foundry-full-course-cu#1510] [Q&A] [answered]
i didn't quite understand why in this test ```testPerformUpkeepRevertsIfCheckUpkeepIsFalse```  rafflestate = 0? Because if ```rafflestate = open```  ```then isOpen is true which makes upKeepNeeded``` also true in checkUpkeep function in  ```Raffle.sol```
OR if possible can someone please explain this whole test because i'm a little confused in this one. Like How ```checkUpkeep``` is false in it?

Raffle.sol

```// SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {console} from "../lib/forge-std/src/console.sol";
// import {VRFCoordinatorV2Interface} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFCoordinatorV2Interface} from "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";
// import {VRFConsumerBaseV2} from "../lib/chainlink-brownie-contracts/contracts/src/v0.8/VRFConsumerBaseV2.sol";
/**
 * @title A sample Raffle contract
 * @author Anshuman
 * @notice This contract is for creating a sample raffle.
 * @dev Implements Chainlink VRFv2
 */
contract Raffle is VRFConsumerBaseV2 {
    error Raffle__NotEnoughEth();
    error Raffle__TransferFailed();
    error Raffle__RaffleNotOpen();
    error Raffle__UpkeepNotNeeded(uint256 currentBalance, uint256 numPlayers, uint256 raffleState);

    enum RaffleState {
        OPEN,
        CALCULATING
    }
    
    uint16 private constant REQUEST_CONFIRMATION = 3;
    uint32 private constant NUM_WORDS=1;
    uint256 private immutable i_entranceFee;
    uint256 private immutable i_interval;
    address private s_recentWinner;
    uint256 private s_lastTimeStamp;
    bytes32 private immutable i_gasLane;
    VRFCoordinatorV2Interface private immutable i_vrfCoordinator;
    uint64 private immutable i_subscriptionId;
    uint32 private immutable i_callBackGasLimit;
    address payable[] private s_player; 
    RaffleState private s_raffleState;

 
    event PickedWinner(address indexed winner);
    event RequestedRaffleWinner(uint256 indexed requestId);
    event EnteredRaffle(address indexed player);

    constructor(
        uint256 entranceFee,
        uint256 interval,
        address vrfCoordinator,
        bytes32 gasLane,
        uint64 subscriptionId,
        uint32 callBackGasLimit
    )
     VRFConsumerBaseV2(vrfCoordinator)
    {

        i_entranceFee = entranceFee;
        i_interval = interval;
         i_vrfCoordinator = VRFCoordinatorV2Interface(vrfCoordinator);
        i_gasLane = gasLane;
        i_subscriptionId = subscriptionId;
        i_callBackGasLimit = callBackGasLimit;

        s_raffleState = RaffleState.OPEN;
        s_lastTimeStamp = block.timestamp;
        
    }

    function  enterRaffle ()  external payable  {
        // require(msg.value >= i_entranceFee,"Not enough eth sent");
        //  console.log("Received Ether:", msg.value);
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEth(); 
        }

        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__RaffleNotOpen();
        }

        s_player.push(payable(msg.sender));
        emit EnteredRaffle(msg.sender);
    }

    /**
     * @dev This is the function that the Chainlink Automation nodes call
     * to see if it's time to perform an upkeep.
     * The following should be true for this to return true:
     * 1. The time interval has passed between raffle runs 
     * 2. The raffle is in open state
     * 3.The contract has eth(aka,player)
     * 4. (Implict) The subscription is funded with LINK
     */
    function checkUpkeep(
        bytes memory /* checkdata */
    ) public view returns (bool upkeepNeeded, bytes memory) {
        bool timeHasPassed =(block.timestamp - s_lastTimeStamp) >= i_interval;
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool hasBalance = address(this).balance > 0;
        bool hasPlayers = s_player.length >0;
        upkeepNeeded = (timeHasPassed && isOpen && hasBalance && hasPlayers);
        return (upkeepNeeded,"0x0");  
    }

    function performUpkeep(bytes calldata) external {
        (bool upkeepNeeded,) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(
                address(this).balance,
                s_player.length,
                uint256(s_raffleState)
            );
        }
        s_raffleState = RaffleState.CALCULATING;
         uint256 requestId = i_vrfCoordinator.requestRandomWords(
            i_gasLane, 
            i_subscriptionId,
            REQUEST_CONFIRMATION,
            i_callBackGasLimit,
            NUM_WORDS
        );
        emit RequestedRaffleWinner(requestId);
    }

    function fulfillRandomWords(
        uint256 /*requestId */ ,
        uint256[] memory randomWords
    )internal override{
        // checks
        // Effects (our own contract)
        uint256 indexOfWinner = randomWords[0] % s_player.length;
        address payable winner = s_player[indexOfWinner];
        s_recentWinner = winner;
        s_raffleState = RaffleState.OPEN;
        s_player = new address payable[](0);
        s_lastTimeStamp = block.timestamp;
        emit PickedWinner(winner);

        //Interactions(other contracts)

        (bool success,) = winner.call{value: address(this).balance}("");
        if(!success){ 
            revert Raffle__TransferFailed();
        }
    
    } 

/** Getter funtion */
    function getEntranceFee() external view returns (uint256){
        return i_entranceFee;
    }

    function getRafflestate() external view returns (RaffleState){
        return s_raffleState;
    }

    function getPlayer(uint256 indexOfPlayer) external view returns(address ){
        return s_player[indexOfPlayer];
    }

    function getRecentWinner() external view returns (address){
        return s_recentWinner;
    }

    function getLengthOfPlayers() external view returns(uint256) {
        return s_player.length;
    } 

    function getLastTimeStamp() external view returns(uint256) {
        return s_lastTimeStamp;
    } 
}
```
In Raffletest.t.sol specifically this function **testPerformUpkeepRevertsIfCheckUpkeepIsFalse**

```  
function testPerformUpkeepRevertsIfCheckUpkeepIsFalse() public {
        uint256 currentBalance = 0;
        uint256 numPlayers = 0;
        uint256 raffleState = 0;

        vm.expectRevert(
            abi.encodeWithSelector(
                Raffle.Raffle__UpkeepNotNeeded.selector,
                currentBalance,
                numPlayers,
                raffleState
            )
        );
        raffle.performUpkeep("");
    }

**Comment by EngrPips** [comment, marked as answer]
Hello @codeplay22 , If i understand properly what you are confused about is how to know if `upKeep` is needed for our contract or not, right ? 

If that is correct then we need to look at the `checkUpkeep` function of our `raffle.sol` contract and understand what it checks and how it conclude if our contract needs `upKeep` or not. so lets take a look at the function below.

```javascript
 /**
     * @dev This is the function that the Chainlink Automation nodes call
     * to see if it's time to perform an upkeep.
     * The following should be true for this to return true:
     * 1. The time interval has passed between raffle runs 
     * 2. The raffle is in open state
     * 3.The contract has eth(aka,player)
     * 4. (Implict) The subscription is funded with LINK
     */
    function checkUpkeep(
        bytes memory /* checkdata */
    ) public view returns (bool upkeepNeeded, bytes memory) {
        bool timeHasPassed =(block.timestamp - s_lastTimeStamp) >= i_interval;
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool hasBalance = address(this).balance > 0;
        bool hasPlayers = s_player.length >0;
        upkeepNeeded = (timeHasPassed && isOpen && hasBalance && hasPlayers);
        return (upkeepNeeded,"0x0");  
    }
```
The `checkUpkeep` above do a total of 4 checks and all of the check have to return true before it can conclude that our `raffle.sol` contract needs `upKeep`, now let's look at each of the conditions it checks for and what they mean one by one.

**FIRST CHECK**

```javascript
bool timeHasPassed =(block.timestamp - s_lastTimeStamp) >= i_interval;
```
The above check is done to confirm that the current  `block.timestamp` subtracted by last block.timestamp `s_lastTimeStamp` is greater than some sort of interval we have pre-defined in our contract. Basically this check is just to confirm if enough time has passed to pick a winner. assume we decide to pick a winner every one week so it is this check that would confirm if one week has passed

**SECOND CHECK**

```javascript
bool isOpen = RaffleState.OPEN == s_raffleState;
```
The above check is done to confirm that the contract is currently in it's open state, what this basically means is that the check confirms that there is not an ongoing `upKeep`. i mean if we are currently performing `upKeep` on our contract then we don't need any `upKeep` at the moment.

**THIRD CHECK**

```javascript
bool hasBalance = address(this).balance > 0;
```
The above check is done to confirm that the ETHER balance of our contract is greater than 0.
**FOURTH CHECK**

```javascript
bool hasPlayers = s_player.length >0;
```
The above check confirm that the length of `s_player` array is greater than 0, that means it check that at least one person have entered the game.

Only and only if all of this condition returns true that is when the `checkUpkeep` function would return true. confirming that our contract needs `upKeep`. This is enforced by the `&` operator used in the below final check 

** RESULT CHECK**

```javascript
upkeepNeeded = (timeHasPassed && isOpen && hasBalance && hasPlayers);
return (upkeepNeeded,"0x0");  
```
If only one of the for check that was done initially by the contract is false then the `upkeepNeeded` would be equal `false` and the `checkUpkeep` function would return the `upkeepNeeded` which would be false meaning that the contract does not need `upKeep`

So after all of this explanation, let's look at the `testPerformUpkeepRevertsIfCheckUpkeepIsFalse` test that you found confusing

```javascript
function testPerformUpkeepRevertsIfCheckUpkeepIsFalse() public {
        uint256 currentBalance = 0;
        uint256 numPlayers = 0;
        uint256 raffleState = 0;

        vm.expectRevert(
            abi.encodeWithSelector(
                Raffle.Raffle__UpkeepNotNeeded.selector,
                currentBalance,
                numPlayers,
                raffleState
            )
        );
        raffle.performUpkeep("");
    }
```

In the test above the conditions of our contracts shows the below;

- currentBalance = 0; This would make The third check in our contract `checkUpkeep` function `bool hasBalance = address(this).balance > 0;` to return false.

- numPlayers = 0; This would make the Fourth check in our contract `checkUpkeep` function `bool hasPlayers = s_player.length >0;` to return false.

- raffleState = 0; This would make the Second check in our contract `checkUpkeep` function  `bool isOpen = RaffleState.OPEN == s_raffleState;` to return true

- Obviously enough time has not pass has well; This would make the First check in our contract   `bool timeHasPassed =(block.timestamp - s_lastTimeStamp) >= i_interval;` to return false

So on doing the final check we would have the below 

```javascript
upkeepNeeded = (false && true && false && false);
return (upkeepNeeded,"0x0");  
```
Since we know that every condition needs to be `true` before our contract `checkUpkeep` function conclude that our contract needs `upKeep` then we know for sure the contract would return confirming that our contract do not need any `upKeep`.
Honestly i really hope i have done enough to clear your confusion. Took mw 25 minutes to write this out lol.

**Reply by codeplay22** [reply]
@EngrPips Thanks a lot for replying and i really appreciate how quickly you responds to my queries but i wanted to know ```why we want to set raffleState to OPEN or 0 or True when we want upKeep to be false? ``` ```why not to set raffleState to 1 or false ?```   , knowing that all conditions need to be true for it to be TRUE, Isn't that an error? @PatrickAlphaC

**Reply by EngrPips** [reply]
By default the `raffleState` would be open unless we explicitly set it to something else

**Reply by codeplay22** [reply]
@EngrPips Aren't we declaring here explicitly? as  ``` raffleState = 0```

**Reply by EngrPips** [reply]
@codeplay22  we are not explicitly declaring `raffleState` on the raffle contract but just declaring a value that we would use in our  foundry `vm.expectRevert` cheat code as shown below 

```javascript
vm.expectRevert(
            abi.encodeWithSelector(
                Raffle.Raffle__UpkeepNotNeeded.selector,
                currentBalance,
                numPlayers,
                raffleState
            )
        );
```

**Reply by codeplay22** [reply]
@EngrPips can you tell me how can we declare ```raffleState explicitly?``` in that test  ```testPerformUpkeepRevertsIfCheckUpkeepIsFalse```

**Reply by EngrPips** [reply]
You change the `raffleState` in the `performUpkeep` function of the `Raffle.sol` contract , read the below code to see where the `raffleState` is altered explicitly.

```javascript
function performUpkeep(bytes calldata) external {
        (bool upkeepNeeded,) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(
                address(this).balance,
                s_player.length,
                uint256(s_raffleState)
            );
        }
        s_raffleState = RaffleState.CALCULATING;
         uint256 requestId = i_vrfCoordinator.requestRandomWords(
            i_gasLane, 
            i_subscriptionId,
            REQUEST_CONFIRMATION,
            i_callBackGasLimit,
            NUM_WORDS
        );
        emit RequestedRaffleWinner(requestId);
    }
```

**Reply by codeplay22** [reply]
@EngrPips is it even executing that line ```s_raffleState = RaffleState.CALCULATING; ? ``` i think the code is reverting back from the ```if condition only From here```     
```
if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(
                address(this).balance,
                s_player.length,
                uint256(s_raffleState)
            );
        }
```
what do you think?

**Reply by EngrPips** [reply]
If upkeep is not needed it would revert before it gets to this line `s_raffleState = RaffleState.CALCULATING;`

**Reply by codeplay22** [reply]
so do we actually need it in the above code?

**Reply by EngrPips** [reply]
Do we need what ?

**Reply by codeplay22** [reply]
i meant upkeep is needed or not? in this test function ```testPerformUpkeepRevertsIfCheckUpkeepIsFalse```

**Reply by EngrPips** [reply]
I am not sure how you mean, but the test function `testPerformUpkeepRevertsIfCheckUpkeepIsFalse` test to confirm that the `Raffle.sol` contract would return false when upkeep is not needed

**Reply by codeplay22** [reply]
i meant when i run this test function ```testPerformUpkeepRevertsIfCheckUpkeepIsFalse``` will this code run or not?
```  
s_raffleState = RaffleState.CALCULATING;
         uint256 requestId = i_vrfCoordinator.requestRandomWords(
            i_gasLane, 
            i_subscriptionId,
            REQUEST_CONFIRMATION,
            i_callBackGasLimit,
            NUM_WORDS
        );
        emit RequestedRaffleWinner(requestId); 
```

**Reply by EngrPips** [reply]
no it will not run

**Reply by codeplay22** [reply]
Thanks 
That's what i was saying too! anyways thank you for explaining. :)

**Reply by EngrPips** [reply]
Oh i see, glad it is clarified
